### DISCUSSION TITLE
How to set the prototype of DynamicObject correctly?

### DISCUSSION THREAD

**Discussion by zllovesuki** [dop251/goja#488] [Q&A] [answered]
On godoc:

*The Object's prototype is initially set to Object.prototype, but can be changed using regular mechanisms (Object.SetPrototype() in Go or Object.setPrototypeOf() in JS).*

*The Object cannot have own Symbol properties, however its prototype can. If you need an iterator support for example, you could create a regular object, set Symbol.iterator on that object and then use it as a prototype. See TestDynamicObjectCustomProto for more details.*

However, I'm having trouble of doing so from Go side.

Suppose that I have a class:
```javascript
class FetchEvent {
    get [Symbol.toStringTag]() {
        return 'FetchEvent'
    }
}
```

Calling `.toString()` on an instance of `FetchEvent` returns `[object FetchEvent]` correctly. but I'm trying to set a DynamicObject's prototype to `FetchEvent`, so I don't get `TypeError: Could not convert &{{0xc00011f380 0xc000735b90} 0xc000b84000} to primitive` when I'm trying to call `.toString()` on the DynamicObject.

What's the correct way of doing so? I have tried:
```javascript
class FetchEvent {
    get [Symbol.toStringTag]() {
        return 'FetchEvent'
    }
}

const instance = new FetchEvent()
```
```golang
	evt.nativeEvt = vm.NewDynamicObject(evt)
	class := vm.Get("FetchEvent")
	instance := vm.Get("instance")
	evt.nativeEvt.SetPrototype(class.ToObject(vm))
	fmt.Printf("%+v\n", evt.nativeEvt)
	evt.nativeEvt.SetPrototype(instance.ToObject(vm))
	fmt.Printf("%+v\n", evt.nativeEvt)
	evt.nativeEvt.SetPrototype(class.ToObject(vm).Prototype())
	fmt.Printf("%+v\n", evt.nativeEvt)
	evt.nativeEvt.SetPrototype(instance.ToObject(vm).Prototype())
	fmt.Printf("%+v\n", evt.nativeEvt)
```

But all of them still fails with `%!v(PANIC=String method: TypeError: Could not convert &{{0xc0003b0f30 0xc0005c5bc0} 0xc000972000} to primitive)`

Maybe I'm just bad at JavaScript 😅

**Comment by zllovesuki** [comment]
I figured it out. Even if this is dynamic, it is still trying to access `.toString` or `.valueOf` from the DynamicObject itself, you can 302 those `.Get()` to the underlying instance itself:

```golang
// ...
	evt.nativeEvt = evt.vm.NewDynamicObject(evt)
	evt.nativeEvt.SetPrototype(evt.nativeEvtInstance)
// ...

func (evt *fetchEvent) Get(key string) goja.Value {
	switch key {
	case "respondWith":
		return evt.vm.ToValue(evt.nativeRespondWith)
	case "waitUntil":
		return evt.vm.ToValue(evt.nativeWaitUntil)
	case "request":
		return evt.requestProxy.nativeReq
	default:
		return evt.nativeEvtInstance.Get(key)
	}
}
```

Now it is no longer exploding 😆
```
2023-03-11T03:46:35.274-0800    INFO    heresy/runtime_logger.go:23     [object FetchEvent]     {"script": "event.js"}
```

**Comment by dop251** [comment, marked as answer]
Yes, I was just writing to suggest that your DynamicObject implementation that either returns a primitive or a function that returns an object when asked for the `toString` property.

The following works as expected for me:

```go
func TestDynamicObjectProtoSymbol1(t *testing.T) {
	vm := New()
	_, err := vm.RunString(`
class FetchEvent {
    get [Symbol.toStringTag]() {
        return 'FetchEvent'
    }
}

const instance = new FetchEvent()
	`)
	if err != nil {
		t.Fatal(err)
	}
	m := make(map[string]Value)
	//m["toString"] = vm.ToValue("")
	dynObj := &testDynObject{
		r: vm,
		m: m,
	}
	o := vm.NewDynamicObject(dynObj)
	class := vm.Get("FetchEvent")
	instance := vm.Get("instance")
	o.SetPrototype(class.ToObject(vm))
	t.Logf("%+v\n", o)
	o.SetPrototype(instance.ToObject(vm))
	t.Logf("%+v\n", o)
	o.SetPrototype(class.ToObject(vm).Prototype())
	t.Logf("%+v\n", o)
	o.SetPrototype(instance.ToObject(vm).Prototype())
	t.Logf("%+v\n", o)
}
```

Produces:

```
    object_dynamic_test.go:487: %!v(PANIC=String method: TypeError: Function.prototype.toString requires that 'this' be a Function)
    object_dynamic_test.go:489: [object FetchEvent]
    object_dynamic_test.go:491: %!v(PANIC=String method: TypeError: Function.prototype.toString requires that 'this' be a Function)
    object_dynamic_test.go:493: [object FetchEvent]
```

(BTW, the second scenario is the right one)

However, if I uncomment the `//m["toString"] = vm.ToValue("")` line, I get the following:

```
    object_dynamic_test.go:487: %!v(PANIC=String method: TypeError: Could not convert &{{0xc000206870 0xc000206180} 0xc0001f8df0} to primitive)
    object_dynamic_test.go:489: %!v(PANIC=String method: TypeError: Could not convert &{{0xc000206870 0xc0002067e0} 0xc0001f8df0} to primitive)
    object_dynamic_test.go:491: %!v(PANIC=String method: TypeError: Could not convert &{{0xc000206870 0xc00010f8c0} 0xc0001f8df0} to primitive)
    object_dynamic_test.go:493: %!v(PANIC=String method: TypeError: Could not convert &{{0xc000206870 0xc000206120} 0xc0001f8df0} to primitive)

```

**Reply by zllovesuki** [reply]
so `o.SetPrototype(instance.ToObject(vm).Prototype())` is the one I should be using, correct? thank you 🙏

**Reply by dop251** [reply]
Yes, or it can be `class.ToObject(vm).Get("prototype").ToObject(vm)`, this way you don't need the instance. Also, `Symbol.toStringTag` can be a simple property, no need to make it a getter.
